### PR TITLE
Jetpack connect: Fix getSiteConnectInfo url handling

### PR DESCRIPTION
**Pull request by sirreal**
Depends on D11731-code

URLs were being passed to the endpoint in a complex and confusing way. Pass them as a GET param to the new endpoint.

Improvements around valid urls and tests have been moved to #24393 

Adds a bunch of tests 🎉 

## Testing
1. Apply D11731-code and sandbox `public-api.wordpress.com`
1. Visit https://calypso.live/jetpack/connect?branch=fix/jpc-site-info-encoding and try a series of URLs. Look at the tests for some good candidates:

https://github.com/Automattic/wp-calypso/blob/c527809b8e4ef21010c21163371170f62e12b64b/client/lib/wpcom-undocumented/lib/test/get-site-connect-info.js#L34-L47

https://github.com/Automattic/wp-calypso/blob/74f3fa12ef95a1dc8dc73a4d82259af662095d4c/client/jetpack-connect/test/utils.js#L34-L40

### CODE REVIEW

**Comment by sirreal** [comment]
Thanks for testing, @tyxla!

It would be great to get D11731-code merged as an incremental improvement. One thing on the roadmap is improving redirection handling and the changes there and in this PR lay some foundations.

I'll take a look at the specific case you mention. I don't think that's a blocker as it's not well handled by the existing implementation and I'd like to make sure these changes are small and incremental.

We do some "cleaning" on the input, but that's handled in the Jetpack connect components. It's a bit of a fuzzy line where that handling should go. Ideally, I think this API call should make sure what we're sending is well formed and adheres to what the API expects. Doing "smarter" things like trimming known `/wp-admin` routes from a provided URL may be good ideas, but I think they should be handled when we treat the input:

https://github.com/Automattic/wp-calypso/blob/cbf2dc5f8ba992ba4751e70c1312906b6ca4e437/client/jetpack-connect/utils.js#L71-L84

**Comment by seear** [comment]
> I'll take a look at the specific case you mention. I don't think that's a blocker as it's not well handled by the existing implementation and I'd like to make sure these changes are small and incremental.

Yeah agree with this. 

I think we can do another PR that simply changes `/wp-admin\/?$/` to `/wp-admin.*$/`

**Comment by seear** [comment]
How about encoding the protocol part as well instead of passing separately? For example:
`http%3A%2F%2Fmodern-copperhead.w5.poopy.life%2F`

Or is that planned for a separate change?

**Comment by sirreal** [comment]
> How about encoding the protocol scheme part as well instead of passing separately?

I think that would be a good change as well. I was initially avoiding it because I didn't want to make breaking changes, but we already are. However, that would change the API signature, which we may be able to work around but is a bit more invasive:

```php
[
  'path' => '/connect/site-info/%s/%s',
]
```

This would need to be `'path'        => '/connect/site-info/%s',`. We could temporarily put some filler on the end so it continues to match `/%s/%s`.

If we go this route, I'd also consider moving everything to the query. I'm not fluent in REST, but we're not reading items or collections that belong on the URL, we're passing a (required) parameter so that an action which will return an ephemeral result can be executed.

Thoughts?

**Comment by seear** [comment]
> that would change the API signature

Yeah, maybe this should be for another iteration.

>  we're not reading items or collections that belong on the URL, we're passing a (required) parameter so that an action which will return an ephemeral result can be executed.

What you say there makes sense. However, from memory, I think this wpcom API mostly takes site URLs in the path, so I'd be tempted to stick with that.

**Comment by sirreal** [comment]
> However, from memory, I think this wpcom API mostly takes site URLs in the path, so I'd be tempted to stick with that.

wpcom API takes site _slugs_ in the path. Those _are_ more traditional requests in the REST sense. A slug is a unique identifiers for a site from the collection of sites. I think that conflating these concepts is one reason this endpoint was implemented in a peculiar way. 

Here, we are passing a url as a parameter to an action that will be run and return some result that represents the current, ephemeral, state of things. This is not an identifier for a site from a collection of sites (at least not in the wpcom sense), but more like an argument for a function.

I'm not convinced moving the URL to a query is the correct thing, so I'd like to hear more thoughts 🙂

**Comment by seear** [comment]
Your logic sounds faultless to me, but at this point I don't think it matters too much, and I genuinely have no preference either way. We did the remote install with the URL on the path and the user and password as queries.

One thing to be said is that changing to a query might make it easier to push these changes through. At the endpoint, if we receive a query arg, treat it as a standard URL. Anything on the path—keep the existing processing. We wouldn't need any versioning at all then, and we can remove the path part of the endpoint when things have settled down.

File: client/lib/wpcom-undocumented/lib/undocumented.js [hunk 1]
```diff
@@ -2193,26 +2192,20 @@ Undocumented.prototype.getExport = function( siteId, exportId, fn ) {
...
-	const parsedUrl = url.parse( targetUrl );
-	let endpointUrl = `/connect/site-info/${ parsedUrl.protocol.slice( 0, -1 ) }/${ parsedUrl.host }`;
-
-	if ( parsedUrl.path && parsedUrl.path !== '/' ) {
-		endpointUrl += parsedUrl.path.replace( /\//g, '::' );
-	}
-
-	return this.wpcom.req.get( endpointUrl );
+Undocumented.prototype.getSiteConnectInfo = function( inputUrl ) {
+	return this.wpcom.req.get( '/connect/site-info', { url: inputUrl } );
```
Comment (sirreal) [inline client/lib/wpcom-undocumented/lib/undocumented.js]: Now that the url moves to the query, we can rely on existing underlying mechanisms to handle escaping ⭐️

**Comment by sirreal** [comment]
Great, thanks @seear!

Labelled as blocked pending D11731-code (pMz3w-8vE-p2)

**Pull request merged**